### ISSUE TITLE
Hashtag spam protection

### ISSUE THREAD

**Issue by hoergen** [friendica/friendica#8540]
Today I got a post from a person I don't follow  and he neither was shared. BUT I got his post via hashtag. That would be okay, If he wouldn't have used every country in this world as hashtag.

![Screenshot_019](https://user-images.githubusercontent.com/24474762/80284531-66674000-871f-11ea-8d1d-a71b26900a67.png)
source: https://social.outsourcedmath.com/display/9b1e8f93-125e-a45c-85c7-4f2891682801

It would be good  if there would be some hashtag spam protection as a server and a user setting (Server over User Setting)  where you can define the maximum number of hashtags that are allowed on this server or on this user.

BTW

The post in my timeline looks like this (without any body)

![Screenshot_020](https://user-images.githubusercontent.com/24474762/80284705-66b40b00-8720-11ea-847b-7518cebe989d.png)

**Comment by AlfredSK** [comment]
You can use the Advanced Content Filter add-on to collapse posts with more than X hashtags.

**Comment by MrPetovan** [comment]
I personally have disabled the tags display because it's overwhelmingly redundant with the message body. What is it used for exactly?

**Comment by AlfredSK** [comment]
Well, users like it. I asked them if I should disable these tag block and the answer was 'please don't'. :-)

They're useful for posts where the tags and mentions are spread over the whole body of the post. In this case it's like a tag summary.

Also, when you use post categories and saved folders they're displayed there as well. And that's pretty cool.

**Comment by MrPetovan** [comment]
That last part is the reason why I said "overwhelmingly".

Unfortunately, the way we do post collapsing is tightly coupled with the post body itself.

**Comment by hoergen** [comment]
Sorry guys this is not about half baked nerd addons like the Advanced Content Filter  that is unusable for a stand user and not about fancy displays. Sure I could use it and have even more collapsed (instead of deleted) posts in my timeline. 

It is about spam protection .

Please investigate before just assuming and  giving some developsplaining answers.
You can see, if you would have visited the original post, that the hashtags are in the post body itself not as the  extra hashtag display.

**Comment by MrPetovan** [comment]
I know the period is hard for everybody, but you do not have to be condescending in your written messages to people you expect to provide free work for your convenience. This isn't about security nor spam protection.

You got one (1) post you didn't like to see in your feed and instead of manually removing it or blocking the contact for repeated offense you came here to ask a bunch of work from us. Then when someone actually provided you with a crowd-sourced workaround, you got pissy at them and the devs.

What do you expect to happen exactly?

**Comment by hoergen** [comment]
I'm just reporting stuff  I'm watching for some times and if I think this could be a problem in future I open an issue for that. If somebody is working on it or not that is a completely other thing. But what is really toxic in here is are comments "This isn't a problem because it happens to you once. You can delete it if you don't like it". 
There are a lot things you can ignore, if they happens only once or only sometimes. Sitting in the edge and covering your eyes won't make them disappear.

So please leave it, if you are not interested and don't spread such toxic comments. And if the project decides not to approach such a door for spam, there is still a "wont fix" label.
Thank you very much.

**Comment by SpcCw** [comment]
I was bothered with similar issue as well. So far it isn't too bad but it might easily get worse.

Advanced Content Filter isn't bad at all and I am pretty sure even non-nerds can use it but it needs a couple of improvements in my opinion.

1. Popular presets. Maybe in the form of rules shipped as disabled but editable (too many tags, too many images...) so all users would have them ready to enable if they want. It takes some time to figure out how to filter posts with multiple images even for moderately nerdy person :)

2. It is probably not easily doable but it would be really nice to be able to completely filter posts from the stream. Not just collapse but completely exclude. So the user could have option to either collapse or exclude. If this causes too much performance hit when selecting posts for the stream maybe it could be done like Hubzilla where blacklisted items just don't get stored at all.

**Comment by MrPetovan** [comment]
You're right about the choice to collapse/hide completely, unfortunately at the moment it is tricky because, unlike many other social networks, we don't have a dedicated data structure for the home timeline. You either receive a post or you don't, no filtering can currently happen at that point. That's why we only have the ability to collapse post, because currently post exclusion would have to happen after pagination and it would result in wildly different page item counts. It's already happening to some extent with blocked/ignored accounts if I remember correctly.

I'm not exactly sure how it works in other software, but from what I gather, it's a two-tiered system:

1. All received posts are stored in a table.
2. Then a background process picks and chooses what posts should actually be part of the home timeline based on user preferences: ignore, blocks and miscellaneous filters.

Then the timeline table is normally queried and paginated for display without "holes". This table can be regenerated at will depending on user preferences update.

**Comment by SpcCw** [comment]
@MrPetovan, yes, this sounds like a challenge.

I don't like the idea of server-wide filtering where posts matching blacklist conditions just get dropped. Well, actually it would be nice to have but it would get very limited and careful use. For example, there is #checkin event running on Fediverse (Diaspora part mainly) and posts have 150+ tags (mentions) which are part of their workflow. At the same time spammers sometimes use just a few popular tags.

I can live with more holes in the stream though as long as it is user choice. Personally I don't care much about precise pagination.

**Comment by MrPetovan** [comment]
Unfortunately it can have unintended consequences. For example infinite scroll pagination ends when a given query returns 0 items because we assume it's the very beginning of the item list. However if we keep adding post-query filtering, a given page in the middle of the timeline could end up empty, which would prevent loading the timeline further.

**Comment by annando** [comment]
I don't think that the actual coding would be that complicate. The question is more how to detect spam. We could enable some hard filters (like the number of hashtags). But then we could end up with different decisions if the person had only subscribed to one of these hashtag or if the person had subscribed to the author who used that much tags.

I'm planning some different way of hashtag subscribing for the future. But I cannot tell when the "post" structure will be ready to ship.

**Comment by tobiasd** [comment]
I would not count the number of hash tags. An essay of 5000 words might contains 20 tags which would be too much (for my taste) in a 500 character posting but ok for that essay. A ration would be more appropriate I think, and it needs to be a user setting.

**Comment by SpcCw** [comment]
I don't think there is direct relation between post length and number of tags. Tags are used for reference so if the post is summary of some kind ("twenty coldest cities in the world"), it makes sense to tag each item and at the same time there can be very few words.

So it is crucial to make anything like this user setting or there will be lots of false-positives.

**Comment by annando** [comment]
Like I said: The configuration is the bigger part of the problem.

**Comment by bkil** [comment]
Spam filtering could also be implemented in different ways.

Standard mail spam filters come to mind that could be ran on the body that each user could train (looking into the statistics of the text and some fix features like links, domains, perhaps even words present on the referenced external site) and maybe users could also share their ratings/databases to collaborate.

Spam scores could be computed for each user account based on its metadata (like age), how many posts it has produced and how spammy they appear to be (based on median/mean/max/percentile), and users could also set a spam score limit to hide offenders when following tags.

Surely this could be implemented in a multi-factor way, similar to how it is done in email - users who we have interacted with before or who other people have interacted before (except for disliking!) could receive a score boost for example.

Then from a research perspective we could actually implement tag taxonomy mining to flag posts based on how their individual tags align or contradict each other or a given subset of that specified.

We could do natural language processing to validate the probability that a given tag does not apply to the text body (it could be easier than determining whether it does). If a given proportion of tags are highly likely to mismatch the body, it could increase its spamminess.

Manual flagging/reporting of posts and accounts as spam could be introduced, and after a given number of trusted flag events (maybe based on transitive web of trust or mining it based on interactions), the post would get hidden until moderator intervention, so the least amount of users could see the given post.

These are all very exciting developments, but they all cost a lot of time (=money), hence YAGNI dictates that we should postpone it until it becomes a problem that we need to deal with.

And also don't forget - one man's business is another's spam. :shrug: :unamused: :thinking:

**Comment by hoergen** [comment]
Solved? In Progress? Can be closed?

**Comment by schmaker** [comment]
Maybe fixed by https://github.com/friendica/friendica/pull/13855 and https://github.com/friendica/friendica/pull/13856 ?

**Comment by bkil** [comment]
This should definitely improve the situation, but not fix it entirely. After implementing similar protection on matrix, a spammer would then post 100 separate messages - each with 1 link and 10 hashtags to game such a limit.

The content ratio check can also be sidestepped by generating totally random characters (later: random words, then random phrases based on Markov chains, etc.).

So this is always a cat & mouse game. As long as the default checks in Friendica work better than on another server that is deployed more widely, spammers probably won't bother to tweak their actions as long as conversion rate is high enough.

**Comment by hoergen** [comment]
Can I close this issue. As far I can tell .. or feel, there was some progress somewhere that minimizes the problem?

**Comment by bkil** [comment]
Okay. Let's open a new issue linking to this one once it starts to be a burden.

**Comment by hoergen** [comment]
ok since nobody linked another issue to this. I close this. If there is another issue open, please use the search function.
